### PR TITLE
initial linode actions

### PR DESCRIPTION
**Pull request by mjhea0**

### CODE REVIEW

**Comment by mjhea0** [comment]
ready @madflojo :)

**Comment by mjhea0** [comment]
https://assembly.com/runbook/bounties/14

**Comment by madflojo** [comment]
I'll pull this down a bit later and check this out.

**Comment by mjhea0** [comment]
What we really need is a "test this monitor/reaction" button not only for us, but for end users. Similar to how Zapier has it. @madflojo It would only be for when a monitor is tied to a reaction. Want me to open a bounty?

File: src/actions/actions/linode-boot/__init__.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def call_linode(redata, jdata, logger):
+    ''' Perform actual call '''
+    url = 'https://api.linode.com/'
+    params = {
+        "api_action": "linode.reboot",
+        "LinodeID": int(redata['data']['linode_id']),
+        "api_key": str(redata['data']['api_key'])
+    }
+    payload = json.dumps(params)
```
Comment (madflojo) [inline src/actions/actions/linode-boot/__init__.py]: Looks like Linodes API expects them to be POST params vs a JSON message.

> Request Format
> The api_action parameter is required on every request. All API request parameters and method parameters can be sent via either GET or POST variables.
>
> Example request:

https://api.linode.com/?api_key=sekrit&api_action=test.echo&foo=bar
NOTE: api.linode.com responds only via https://api.linode.com.

**Comment by madflojo** [comment]
Agree we should have a test "recipe" button. Go for it on creating a bounty for that.

**Comment by mjhea0** [comment]
updated

**Comment by madflojo** [comment]
Ok, it's working for me now!

The only suggestion I have for this is it seems Linode always replies with a 200-300 response code and you have to parse their JSON messages. It would be nice to have their Error response added to the logger output for debugging.

The above is a nice to have and we can add it later if you want. Also seems like you might need to pull down the latest from `upstream`.

**Comment by mjhea0** [comment]
Can you test again with the `linode-boot` reaction - both a success and a failure. Once done, I'll update the remaining reactions. Thanks, @madflojo !

File: src/actions/actions/linode-boot/__init__.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    try:
+        req = requests.post(
+            url, timeout=3.0, data=params, verify=True)
+    except:
+        return False
+    if req.status_code >= 200 and req.status_code < 300:
+        if len(req.json['ERRORARRAY']) > 0:
+            try:
+                error_message = str(req.json()['ERRORARRAY'][0]['ERRORMESSAGE'])
```
Comment (madflojo) [inline src/actions/actions/linode-boot/__init__.py]: actioner_1      | 2015-01-06 19:24:56,570 - actions.actioner - ERROR - Got error when attempting to run reaction linode-boot for monitor a3c548b1-42ce-45b5-b199-2362f201aa0a
actioner_1      | 2015-01-06 19:24:56,571 - actions.actioner - ERROR - 'instancemethod' object has no attribute '__getitem__'

I'm thinking maybe I didn't get a JSON message from linode? I did this by using an invalid node id (old)

Comment (mjhea0) [inline src/actions/actions/linode-boot/__init__.py]: try now. silly error on my end.

File: src/actions/actions/linode-boot/__init__.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    try:
+        req = requests.post(
+            url, timeout=3.0, data=params, verify=True)
+    except:
+        return False
+    if req.status_code >= 200 and req.status_code < 300:
+        if len(req.json()['ERRORARRAY']) > 0:
+            try:
+                error_message = str(req.json()['ERRORARRAY'][0]['ERRORMESSAGE'])
```
Comment (madflojo) [inline src/actions/actions/linode-boot/__init__.py]:
```
actioner_1      | 2015-01-06 21:11:05,610 - actions.actioner - ERROR - Got error when attempting to run reaction linode-boot for monitor a3c548b1-42ce-45b5-b199-2362f201aa0a
actioner_1      | 2015-01-06 21:11:05,611 - actions.actioner - ERROR - tuple index out of range
```

To duplicate this you can probably just fake the api key and a random number.

**Comment by madflojo** [comment]
Ready for another test?

**Comment by mjhea0** [comment]
yes

**Comment by madflojo** [comment]
Ok I'll pull it down in a bit and give it a whirl. If all goes well you think it will be ready for prod later today? There are two PR's I want to roll into production today, if this is ready I can roll it as well.

**Comment by mjhea0** [comment]
It's ready, but I think I'd like to add the max_limit field. Do you want that included before it goes to prod?

**Comment by madflojo** [comment]
I would say that a max_limit field is necessary for a create reaction but for reboot, boot, and shutdown I think it is a nice to have. We could add it later, or we could release it later too.

**Comment by mjhea0** [comment]
Well, let me know how the test goes. Then we can make that decsion.

**Comment by madflojo** [comment]
Works (errors) like a charm now! I'm good adding the max_limit later if you are.

**Comment by mjhea0** [comment]
Sweet. I'll open a new bounty for the max limit.

**Comment by madflojo** [comment]
Cool, merged!

**Pull request merged**